### DISCUSSION TITLE
[permission] allow async validation

### DISCUSSION THREAD

**Discussion by lewoudar** [eadwinCode/django-ninja-extra#57] [Ideas]
Hi folks,
First of all, **many thanks** to the author of this library. It was the evolution of [DRF](https://www.django-rest-framework.org/) I want to see!
Now, regarding permissions. I recently have a case where I need to call an external API for validation and I thought it would be good to allow asynchronous operation in the Permission class, I mean, something like the following:

```python
class ReadOnly(permissions.AsyncBasePermission):
    async def has_permission(self, request, view):
        # make an asynchronous call
        await httpx.get(...)
        return request.method in permissions.SAFE_METHODS
```
Edit: I guess we will need a new base class for this feature and it will also apply to the method `has_object_permission`
I'm not sure if it is easy to add. What do you think? 😅

**Comment by eadwinCode** [comment]
@lewoudar this is a good idea. I think your example makes a sense. Having a new base class for the `AsyncBasePermission` is better. The problem lie in execution process. Trying to figure out how best to permission sync and async permissions when handling request. And also if merging Auth and Permission is the better to go

**Reply by lewoudar** [reply]
I don't know if merging auth and permission is a good idea (or not). Maybe keep them separate to help the migration of DRF users? 😅

**Reply by eadwinCode** [reply]
Yeah... You have a good point here😅

**Comment by JonasTischer** [comment]
What is the current status of this? How would you @eadwinCode recommend to implement Async Permission checks at the moment?

**Comment by eadwinCode** [comment]
@JonasTischer I totally forgot about this feature. I will create an issue for this and we can track it from there.

Thanks for bringing this to my notice

**Reply by eadwinCode** [reply]
https://github.com/eadwinCode/django-ninja-extra/issues/248 This closes this discussion
